build: Remove `--enable-threadlocal`

fanquake opened the pull request:
Includes a commit from #30099.
Closes #29952.

hebasto committed:
build: Enable `thread_local` for MinGW-w64 builds

The assumption in the commit 188ca75e5fe4837d16241446558c7566912f67b2
about the broken `thread_local` implementation in GCC was misguided
because the initial failure was not due to GCC, but a bug in the Wine
runtime, as evidenced, for example, here:
 - https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=917307
 - https://bugs.freedesktop.org/show_bug.cgi?id=108662

Consequently, it is safe to re-enable `thread_local` support for
MinGW-w64 builds.

fanquake committed:
build: remove --enable-threadlocal

laanwj commented:
Concept ACK.

hebasto commented:
Concept ACK.

theuni reviewed (commented):
Concept ACK

hebasto commented:
> Concept ACK.

I forgot to mention that it would be great to ditch it before CMake :)

fanquake committed:
doc: remove comment about using thread_local

Followup to https://github.com/bitcoin/bitcoin/pull/30095#discussion_r1605655974.

laanwj commented:
Code review ACK 17fe948cce2eb75f0f3f4b0db9d0d90648c7d4af
This removes, to my knowledge, all mentions of thread-local in the build system or comments.
Will test the guix binary on windows.

hebasto reviewed (approved):
ACK 17fe948cce2eb75f0f3f4b0db9d0d90648c7d4af.

Guix build:
```
x86_64
b5dd1ec801cb549590bb922867b9254de51c3688220ce4d29ae03077e0047448  guix-build-17fe948cce2e/output/dist-archive/bitcoin-17fe948cce2e.tar.gz
9bc0b56bf0440f04937646ded534e47df860c4ca2b9b96699c1f5b431ae71508  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/SHA256SUMS.part
3e1cceb01e86b7e47d68ae100b60747333a88aad5b352c7c3e9113d700465f1c  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64-debug.zip
e0c36335cfc1c399d40231adffd1d8afd33261dc59c0523823c50ab6ee00f5e8  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64-setup-unsigned.exe
8a0044e4e5357e4e0eb1577a704a4cc0ddb46617151eee607842010c120738a8  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64-unsigned.tar.gz
de79709714a39ebcb5988fad5add91e7ed71e9fc313143f0cca5dcc6814dd34b  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64.zip
```

fanquake commented:
Guix Build (aarch64):
```bash
77a67096a2225cafa44637d69e9618cb275b26eeda00585dfa246dfbec7fc818  guix-build-17fe948cce2e/output/aarch64-linux-gnu/SHA256SUMS.part
57867960d6c8597648cc49efe4165e543b1181ad0a3a97a16f66bd41b93afe8b  guix-build-17fe948cce2e/output/aarch64-linux-gnu/bitcoin-17fe948cce2e-aarch64-linux-gnu-debug.tar.gz
76c872bcb98e8fc46a5cb2ebd13c489caa6b7d3b0b6db92c467aae27b30308fc  guix-build-17fe948cce2e/output/aarch64-linux-gnu/bitcoin-17fe948cce2e-aarch64-linux-gnu.tar.gz
5093f7155c56e94ad994dcbe586fef1b8cfedc214e037a4dfdb1b2012059b5c9  guix-build-17fe948cce2e/output/arm-linux-gnueabihf/SHA256SUMS.part
06f2bb56d6aae8b1dc70e5d09a111739a7d34b8b07797f674f6c17fbf7ec74ab  guix-build-17fe948cce2e/output/arm-linux-gnueabihf/bitcoin-17fe948cce2e-arm-linux-gnueabihf-debug.tar.gz
83a5f2bcb1ecd4e7cf1cc85db394396aaf46b92252555f42237a44cde0bc9b96  guix-build-17fe948cce2e/output/arm-linux-gnueabihf/bitcoin-17fe948cce2e-arm-linux-gnueabihf.tar.gz
13c6c28083906e8ac81d4f320a6d4d9f7508919530d33f371596baaea8eba066  guix-build-17fe948cce2e/output/arm64-apple-darwin/SHA256SUMS.part
0dcf263fb0c3945af210977dd5a363e3cff162c4a7b47d8fed3cef4b708b9365  guix-build-17fe948cce2e/output/arm64-apple-darwin/bitcoin-17fe948cce2e-arm64-apple-darwin-unsigned.tar.gz
3c9da1723c61bcab2e6a856f0a7cbb3ad53cea7b18b8761b1d50d0cfd3012305  guix-build-17fe948cce2e/output/arm64-apple-darwin/bitcoin-17fe948cce2e-arm64-apple-darwin-unsigned.zip
7b6debab244c7c50e6eb47c180bc149de511f8b2304870c5e2347fdac3107fa1  guix-build-17fe948cce2e/output/arm64-apple-darwin/bitcoin-17fe948cce2e-arm64-apple-darwin.tar.gz
b5dd1ec801cb549590bb922867b9254de51c3688220ce4d29ae03077e0047448  guix-build-17fe948cce2e/output/dist-archive/bitcoin-17fe948cce2e.tar.gz
5743e13fa3b8505a66412f1845e8b0d8f908822b417b4d94412fb381b30db64a  guix-build-17fe948cce2e/output/powerpc64-linux-gnu/SHA256SUMS.part
c394fe9822cba9275d17dc040772325e8c4ae091b3dc4137a3b3856af9358028  guix-build-17fe948cce2e/output/powerpc64-linux-gnu/bitcoin-17fe948cce2e-powerpc64-linux-gnu-debug.tar.gz
5d5ed565449d736520ee8d8d2f15bb55c99f6bea2b5a8bebc30aca75055f67c7  guix-build-17fe948cce2e/output/powerpc64-linux-gnu/bitcoin-17fe948cce2e-powerpc64-linux-gnu.tar.gz
f5565a2e56d65c262f561b9ea509afd22f4d6ff55907cc259fe114dd6880f7d8  guix-build-17fe948cce2e/output/riscv64-linux-gnu/SHA256SUMS.part
8821cd02599d6d26858d7414147ad93e8c7c9f6ff272fde44162dc291332f3ad  guix-build-17fe948cce2e/output/riscv64-linux-gnu/bitcoin-17fe948cce2e-riscv64-linux-gnu-debug.tar.gz
ea4844a3048eda3c31339754ef42efc30a5b2def1b776ffab990add0c523e6a2  guix-build-17fe948cce2e/output/riscv64-linux-gnu/bitcoin-17fe948cce2e-riscv64-linux-gnu.tar.gz
cd0e7b8739c7039205dfc714df6889646a4f5bb55e0cb1819797990bf96fc8f0  guix-build-17fe948cce2e/output/x86_64-apple-darwin/SHA256SUMS.part
d680caba616d24ef7ec28a7100d093ae428908d1b12ea27352fb575269a49c87  guix-build-17fe948cce2e/output/x86_64-apple-darwin/bitcoin-17fe948cce2e-x86_64-apple-darwin-unsigned.tar.gz
fb234d002cb3c68e7bba82cbb10677e4473f24a617f09e2c17f0b69dd2c47fa3  guix-build-17fe948cce2e/output/x86_64-apple-darwin/bitcoin-17fe948cce2e-x86_64-apple-darwin-unsigned.zip
15c310bef26654d0a9a7098a54593fb58266d7de2d5be28a307d98e64716fe83  guix-build-17fe948cce2e/output/x86_64-apple-darwin/bitcoin-17fe948cce2e-x86_64-apple-darwin.tar.gz
4ba52b40fad3b0ccb3a4c9a812a5dc4aab0047b2a908dd18f58bf417a71c60e3  guix-build-17fe948cce2e/output/x86_64-linux-gnu/SHA256SUMS.part
f5eba6a58f87896795928c5010cbee34453b4fa223e37a6b0cb7da65327f8afd  guix-build-17fe948cce2e/output/x86_64-linux-gnu/bitcoin-17fe948cce2e-x86_64-linux-gnu-debug.tar.gz
af9816b48a68116a61ba2e2af1a22f5aa508b7f699acecfb80b31ef8da976c69  guix-build-17fe948cce2e/output/x86_64-linux-gnu/bitcoin-17fe948cce2e-x86_64-linux-gnu.tar.gz
9bc0b56bf0440f04937646ded534e47df860c4ca2b9b96699c1f5b431ae71508  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/SHA256SUMS.part
3e1cceb01e86b7e47d68ae100b60747333a88aad5b352c7c3e9113d700465f1c  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64-debug.zip
e0c36335cfc1c399d40231adffd1d8afd33261dc59c0523823c50ab6ee00f5e8  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64-setup-unsigned.exe
8a0044e4e5357e4e0eb1577a704a4cc0ddb46617151eee607842010c120738a8  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64-unsigned.tar.gz
de79709714a39ebcb5988fad5add91e7ed71e9fc313143f0cca5dcc6814dd34b  guix-build-17fe948cce2e/output/x86_64-w64-mingw32/bitcoin-17fe948cce2e-win64.zip
```

theuni reviewed (approved):
utACK 17fe948cce2eb75f0f3f4b0db9d0d90648c7d4af

vasild commented on src/test/util_threadnames_tests.cpp:
```diff
@@ -52,11 +50,6 @@ std::set<std::string> RenameEnMasse(int num_threads)
-    return;
-#endif
-
     std::set<std::string> names = RenameEnMasse(100);
```
With this it will be possible to remove `export NOWARN_CXXFLAGS="-Wno-error=unreachable-code"` from dev environment on FreeBSD, thanks!

vasild reviewed (commented):
ACK 17fe948cce2eb75f0f3f4b0db9d0d90648c7d4af

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/214 and https://github.com/hebasto/bitcoin/pull/220.
